sympy/solvers/solveset.py

```diff
@@ -931,6 +932,9 @@ def solveset(f, symbol=None, domain=S.Complexes):
             raise NotImplementedError(filldedent('''
                 relationship between value and 0 is unknown: %s''' % b))
 
+    if f.has(Max) or f.has(Min):
+        f = f.rewrite(Piecewise())
```

Comment:
You can just do `f = f.rewrite(Piecewise)`

Reply:
I tried the above change, but some of the tests in `sympy\solvers\tests\test_solveset.py` failed.

Failing tests:
`test_issue_9565`, `test_issue_10069`, `test_issue_9849`, `test_issue_9849 `, `test_invert_real`,`test_solve_abs`, `test_conditionset`, `test_issue_5132_2`

Looks like expressions with `abs()` donot work with this change.

Reply:
On running the test case individually, correct result is obtained. But, still the tests fail for this change. Can't get the reason why?
For [test_issue_9565()](https://github.com/sympy/sympy/blob/master/sympy/solvers/tests/test_solveset.py#L600)
```
>>>print(solveset(Abs((x - 1)/(x - 5)) <=  S(1)/3, x,S.Reals))         # test_issue_9565()
Interval(-1, 2) 
```
But while running tests `Type error` is raised: `TypeError: expecting Expr but got: ExprCondPair`
```
________________________________________________________________________________
_____________ sympy\solvers\tests\test_solveset.py:test_issue_9565 _____________
  File "d:\new\sympy\sympy\solvers\tests\test_solveset.py", line 601, in test_is                                                                                                                sue_9565
    assert print(solveset_real(Abs((x - 1)/(x - 5)) <= S(1)/3, x)) == Interval(-                                                                                                                1, 2)
  File "D:\new\sympy\sympy\solvers\solveset.py", line 957, in solveset_real
    return solveset(f, symbol, S.Reals)
  File "D:\new\sympy\sympy\solvers\solveset.py", line 948, in solveset
    f, symbol, domain=domain, relational=False)
  File "D:\new\sympy\sympy\solvers\inequalities.py", line 483, in solve_univaria                                                                                                                te_inequality
    period = periodicity(e, gen)
  File "D:\new\sympy\sympy\calculus\util.py", line 473, in periodicity
    g_s = decompogen(f, symbol)
  File "D:\new\sympy\sympy\solvers\decompogen.py", line 47, in decompogen
    result += [f.subs(f.args[0], symbol)] + decompogen(f.args[0], symbol)
  File "D:\new\sympy\sympy\solvers\decompogen.py", line 37, in decompogen
    raise TypeError('expecting Expr but got: `%s`' % func_name(f))
TypeError: expecting Expr but got: `ExprCondPair`
 
```

Reply:
Do you mean changing `Piecewise` to `Piecewise()` or including other functions like `Abs`?

Reply:
@ylemkimon I meant including other functions(like `abs()`).
I had removed the `if ` condition. Then the tests failed

Previous code:
```
    else:
            raise NotImplementedError(filldedent('''
                relationship between value and 0 is unknown: %s''' % b))

    if f.has(Max) or f.has(Min):
        f = f.rewrite(Piecewise())

    if isinstance(f, Eq):
```
I tried by removing the `if` condition.
```
    else:
            raise NotImplementedError(filldedent('''
                relationship between value and 0 is unknown: %s''' % b))
 
    f = f.rewrite(Piecewise())

    if isinstance(f, Eq):
```
Is there any difference between `Piecewise` and `Piecewise()`? Although it doesn't seem to be while running this code. I am getting the same results for both.

Reply:
`Abs` has its own solve method `_solve_abs`. So you can leave out `Abs`.

Reply:
`Piecewise()` is an instance of `Piecewise` class, but `rewrites` automatically uses its class. So there is no difference in the result and an instantiation is unnecessary.